src/rmc-compiler/src/main.rs

```diff
@@ -0,0 +1,223 @@
+                .help("Restrict the targets of virtual table function pointer calls."),
+        )
+        .arg(
+            Arg::with_name("sysroot")
+                .long("--sysroot")
+                .help("Override the system root.")
+                .long_help(
+                    "The \"sysroot\" is the location where RMC will look for the Rust \
+                distribution.",
+                ),
+        )
+        .arg(
+            // TODO: Move this to a cargo wrapper. This should return rmc version.
+            Arg::with_name("rustc-version")
+                .short("V")
+                .long("--version")
+                .help("Gets underlying rustc version."),
+        )
+        .arg(
+            Arg::with_name("rustc-options")
+                .help("Arguments to be passed down to rustc.")
+                .multiple(true)
+                .takes_value(true),
+        )
+}
+
+/// Main function. Configure arguments and run the compiler.
+fn main() -> Result<(), &'static str> {
+    let matches = parser().get_matches();
+
+    // Initialize the logger.
+    init_env_logger("RMC_LOG");
+
+    // Generate rustc args.
+    let rustc_args = generate_rustc_args(&matches);
+
+    // Configure queries.
+    let mut queries = QueryDb::default();
+    queries
+        .set_symbol_table_passes(matches.values_of_lossy("symbol-table-passes").unwrap_or(vec![]));
+    queries.set_emit_vtable_restrictions(matches.is_present("restrict-vtable-fn-ptrs"));
+
+    // Configure and run compiler.
+    let mut callbacks = RmcCallbacks {};
+    install_ice_hook();
+    let mut compiler = RunCompiler::new(&rustc_args, &mut callbacks);
+    if matches.is_present("goto-c") {
+        compiler.set_make_codegen_backend(Some(Box::new(move |_cfg| {
+            rustc_codegen_rmc::GotocCodegenBackend::new(&Rc::new(queries))
+        })));
+    }
+    compiler.run().or(Err("Failed to compile crate."))
+}
+
+/// Empty struct since we don't support any callbacks yet.
+struct RmcCallbacks {}
+
+/// Use default function implementations.
+impl Callbacks for RmcCallbacks {}
+
+/// Generate the arguments to pass to rustc_driver.
+fn generate_rustc_args(args: &ArgMatches) -> Vec<String> {
+    let mut gotoc_args =
+        rustc_gotoc_flags(&args.value_of("rmc-lib").unwrap_or(std::env!("RMC_LIB_PATH")));
+    let mut rustc_args = vec![String::from("rustc")];
+    if args.is_present("goto-c") {
+        rustc_args.append(&mut gotoc_args);
+    }
+
+    if args.is_present("rustc-version") {
+        rustc_args.push(String::from("--version"))
+    }
+
+    rustc_args.append(
+        &mut args
+            .values_of("rustc-options")
+            .unwrap_or(clap::Values::default())
+            .map(|s| s.to_string())
+            .collect::<Vec<_>>(),
+    );
+    let sysroot = sysroot_path(args.value_of("sysroot")).unwrap();
+    rustc_args.push(String::from("--sysroot"));
+    rustc_args.push(sysroot.to_string_lossy().to_string());
```

Comment:
Is `to_string_lossy` what we want here? If there are non-valid characters, shouldn't we just fail?

Reply:
Sure. We can fail.